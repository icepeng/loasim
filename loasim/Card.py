from typing import List, Tuple
from loasim.CharacterModifier import CharacterModifier


class Card:
    def __init__(self, name: str, mdf: CharacterModifier) -> None:
        self.name = name
        self.mdf = mdf

    def get_modifier(self):
        return self.mdf


card_list = [
    Card("남겨진 바람의 절벽 (12)", CharacterModifier(crit=7)),
    Card("세상을 구하는 빛 (18)", CharacterModifier(pdamage_indep=7)),
    Card("세상을 구하는 빛 (30)", CharacterModifier(pdamage_indep=15)),
]

card_dict = {card.name: card for card in card_list}


class CardManager:
    def __init__(self, *equipped_list: List[str]) -> None:
        self.equipped_list = equipped_list

    def get_static_modifier(self):
        mdf = CharacterModifier()
        for name in self.equipped_list:
            card = card_dict.get(name)
            if card is None:
                raise Exception(f"{name} is not an card")
            mdf = mdf + card.get_modifier()
        return mdf
