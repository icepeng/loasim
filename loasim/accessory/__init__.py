from loasim.accessory.builtin import lostark_accessory_repository
from loasim.accessory.base import AccessoryState
