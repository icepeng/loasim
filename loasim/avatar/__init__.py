from loasim.avatar.base import AvatarStatus
from loasim.avatar.builtin import lostark_avatar_repository
