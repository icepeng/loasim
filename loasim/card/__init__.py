from loasim.card.builtin import lostark_default_card_repository