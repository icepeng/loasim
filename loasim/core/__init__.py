from loasim.core.buff import (
    AbstractBuff,
    BuffManager,
    BuffState,
    InternalStatOnoffBuff,
    OnoffBuff,
    SkillBuff,
    StackBuff,
    StatBuff,
    StaticBuff,
)
from loasim.core.enemy import Enemy
from loasim.core.skill import AttackPosition, Skill, SkillType
from loasim.core.stat import InternalStat, Stat
