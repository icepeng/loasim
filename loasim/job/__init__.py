from loasim.job.artist import lostark_job_artist
from loasim.job.base import Job, SkillSpecification, SkillState, Tripod
from loasim.job.blade import lostark_job_blade
from loasim.job.hawkeye import lostark_job_hawkeye
from loasim.job.scouter import lostark_job_scouter

job_dict = {
    "artist": lostark_job_artist,
    "blade": lostark_job_blade,
    "hawkeye": lostark_job_hawkeye,
    "scouter": lostark_job_scouter,
}


def get_job(job_name: str) -> Job:
    job = job_dict[job_name]
    if job is None:
        raise ValueError(f"{job_name} is not supported")
    return job
