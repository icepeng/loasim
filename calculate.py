from collections import defaultdict
from typing import List
from loasim.core import Skill


def calculate(dealcycle: List[Skill], cycle_time, enemy):
    total_damage = 0
    damage_dict = defaultdict(float)
    for sk in dealcycle:
        damage = sk.get_damage(enemy)
        total_damage += damage
        damage_dict[sk.name] += damage

    for sk, dmg in damage_dict.items():
        print(sk, f"{dmg / total_damage *100:,.3f}%")
    print(f"total damage: {total_damage:,.0f}")
    print(f"dps: {total_damage / cycle_time:,.0f}")
