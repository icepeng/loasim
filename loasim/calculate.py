from loguru import logger

from collections import defaultdict
from typing import Any, Callable, List, Dict, Literal, Tuple

from pydantic import BaseModel

from loasim.core import Skill, Enemy, BuffManager, InternalStat
from loasim.core.buff import Buff
from loasim.core.stat import Stat
from loasim.setitem import lostark_setitem_repository
from loasim.card import lostark_default_card_repository
from loasim.engraving import lostark_engraving_repository


class DealCycle(BaseModel):
    buff_list: List[Buff]
    skill_list: List[Tuple[Skill, Dict[str, Any], Literal["back", "head"] | None]]
    cycle_time: float


def calculate(
    internal_stat: InternalStat,
    weapon_pdamage: float,
    setitem_state: List[Tuple[str, int, int]],
    engraving_state: List[Tuple[str, int]],
    card_state: List[str],
    generator: Callable[[InternalStat], DealCycle],
    enemy: Enemy,
):
    deal_cycle = generator(internal_stat)
    base_stat = internal_stat.get_stat()
    weapon_stat = Stat(pdamage=weapon_pdamage)  # 무기 품질
    basis_stat = (
        base_stat
        + weapon_stat
        + lostark_setitem_repository.get_stat(setitem_state)
        + lostark_default_card_repository.get_stat(card_state)
    )

    logger.info(base_stat)
    logger.info(basis_stat)

    engraving_buffs = lostark_engraving_repository.get_buffs(engraving_state)
    buff_manager = BuffManager(engraving_buffs + deal_cycle.buff_list)

    total_damage = 0.0
    damage_dict: defaultdict[str, float] = defaultdict(float)
    for skill, buff_state, backhead in deal_cycle.skill_list:
        buff_stat = buff_manager.get_stat(buff_state, skill)
        damage = skill.get_damage(enemy, backhead, basis_stat + buff_stat)
        total_damage += damage
        damage_dict[skill.name] += damage

    for sk, dmg in damage_dict.items():
        print(sk, f"{dmg / total_damage *100:,.3f}%")
    print(f"total damage: {total_damage:,.0f}")
    print(deal_cycle.cycle_time)
    print(f"dps: {total_damage / deal_cycle.cycle_time:,.0f}")
