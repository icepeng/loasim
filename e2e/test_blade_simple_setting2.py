from loasim.calculate import DealCycle, calculate
from loasim.core import BuffState, Enemy, InternalStat
from loasim.job import SkillState


def calculate_blade():
    burst_state = {
        "약점 노출": BuffState(onoff=True),
        "블레이드 아츠": BuffState(onoff=True),
        "버스트": BuffState(onoff=True, stack=12),
    }
    blade_arts_state = {
        "약점 노출": BuffState(onoff=True),
        "블레이드 아츠": BuffState(onoff=True),
        "버스트": BuffState(onoff=False),
    }
    no_blade_arts_state = {
        "약점 노출": BuffState(onoff=True),
        "블레이드 아츠": BuffState(onoff=False),
        "버스트": BuffState(onoff=False),
    }

    deal_cycle: DealCycle = [
        ("스핀 커터", blade_arts_state, "back"),
        ("스핀 커터", blade_arts_state, "back"),
        ("스핀 커터", blade_arts_state, "back"),
        ("윈드 컷", blade_arts_state, "back"),
        ("소울 앱소버", blade_arts_state, "back"),
        ("어스 슬래쉬", blade_arts_state, "back"),
        ("블리츠 러시", blade_arts_state, "back"),
        ("윈드 컷", blade_arts_state, "back"),
        ("블레이드 버스트", burst_state, "back"),
        ("보이드 스트라이크", no_blade_arts_state, "back"),
    ]

    return calculate(
        job_name="blade",
        internal_stat=InternalStat(
            weapon_att=41814, stat_main=137148, crit=568, special=1733, swift=28
        ),
        weapon_pdamage=30,
        setitem_state=[
            ("사멸", 6, 4),
        ],
        engraving_state=[
            ("원한", 3),
            ("슈퍼 차지", 3),
            ("기습의 대가", 3),
            ("예리한 둔기", 3),
            ("버스트", 3),
        ],
        card_state=["남겨진 바람의 절벽 (12)"],
        skill_state={
            "스핀 커터": SkillState(
                level=10,
                tripod={
                    "약점 노출": 1,
                },
            ),
            "소울 앱소버": SkillState(
                level=11,
                gem=7,
                tripod={
                    "암흑 공격": 5,
                    "반 가르기": 4,
                },
            ),
            "보이드 스트라이크": SkillState(
                level=11,
                gem=7,
                tripod={
                    "암흑 공격": 4,
                    "블랙 익스플로젼": 4,
                },
            ),
            "블리츠 러시": SkillState(
                level=11,
                gem=7,
                tripod={
                    "급소 타격": 4,
                    "차지 강화": 5,
                    "쉐도우 러시": 5,
                },
            ),
            "윈드 컷": SkillState(
                level=11,
                gem=5,
                tripod={
                    "지속력 강화": 4,
                },
            ),
            "어스 슬래쉬": SkillState(
                level=10,
                gem=5,
                tripod={
                    "밀어내기": 4,
                    "약점 포착": 4,
                    "대지 폭발": 4,
                },
            ),
            "블레이드 버스트": SkillState(
                level=11,
                gem=7,
            ),
        },
        enemy=Enemy(armor=6000, reduction=23),
        deal_cycle=deal_cycle,
        cycle_time=9.375,
    )


def test_blade_simple_setting2(snapshot: float):
    assert calculate_blade() == snapshot


if __name__ == "__main__":
    calculate_blade()
