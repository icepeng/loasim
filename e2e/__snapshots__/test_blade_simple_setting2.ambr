# name: test_blade_simple_setting2
  7757167.733140017
---
